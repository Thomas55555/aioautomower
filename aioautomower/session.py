"""Module to connect to Automower with websocket."""
import asyncio
import contextlib
import logging
import time
from typing import Literal

from . import rest
from .auth import AbstractAuth
from .const import MARGIN_TIME, MIN_SLEEP_TIME, REST_POLL_CYCLE
from .exceptions import NoDataAvailableException
from .model import HeadlightModes, MowerList

_LOGGER = logging.getLogger(__name__)


class AutomowerEndpoint:
    """Endpoint URLs for the AutomowerConnect API."""

    mowers = "mowers/"
    "List data for all mowers linked to a user."

    actions = "mowers/{mower_id}/actions"
    "Accepts actions to control a mower linked to a user."

    calendar = "mowers/{mower_id}/calendar"
    "Update the calendar on the mower."

    settings = "mowers/{mower_id}/settings"
    "Update the settings on the mower."

    stay_out_zones = "mowers/{mower_id}/stayOutZones/{stay_out_id}"
    "Enabe or disable the stay-out zone."

    work_area_calendar = "mowers/{mower_id}/workAreas{work_area_id}/calendar"
    "Update the calendar for a work area on the mower."


class AutomowerSession:
    """Session."""

    def __init__(
        self,
        auth: AbstractAuth,
        poll=False,
    ) -> None:
        """Create a session.

        :param str api_key: A 36 digit api key.
        :param dict token: A token as returned by rest.GetAccessToken.async_get_access_token()
        :param float ws_heartbeat_interval: Periodicity of keep-alive pings on the websocket in seconds.
        :param loop: Event-loop for task execution. If None, the event loop in the current OS thread is used.
        """
        self.auth = auth
        self.poll = poll
        self.data_update_cbs = []
        self.token_update_cbs = []
        self.rest_task = False
        self.loop = asyncio.get_event_loop()
        self.token = None
        self.data = {}
        self.mowers = {}
        self.listen_task = None
        self.ws_task = None

        self.token_task = None
        self.rest_task = None

    def register_data_callback(self, callback):
        """Register a data update callback.

        :param func callback: Callback fired on data updates. Takes one dict argument which is the up-to-date mower data list.
        :param bool schedule_immediately: Schedule callback immediately (if data is available).
        """
        if callback not in self.data_update_cbs:
            self.data_update_cbs.append(callback)

    def _schedule_data_callback(self, cb, delay=0.0):
        if self.poll:
            if self.data is None:
                raise NoDataAvailableException
        self.loop.call_later(delay, cb, self.mowers)

    def _schedule_data_callbacks(self):
        for cb in self.data_update_cbs:
            self._schedule_data_callback(cb)

    def unregister_data_callback(self, callback):
        """Unregister a data update callback.

        :param func callback: Callback fired on data updates. Takes one dict argument which is the up-to-date mower data list.
        """
        if callback in self.data_update_cbs:
            self.data_update_cbs.remove(callback)

    async def connect(self):
        """Connect to the API.

        This method handles the login and starts a task that keep the access
        token constantly fresh. Also a REST taks will be started, which
        periodically polls the REST endpoint. This method works only, if the
        token is created with the Authorization Code Grant. Call this method
        before any other methods.
        """

        self._schedule_data_callbacks()

        if self.poll:
            await self.get_status()
            self.rest_task = self.loop.create_task(self._rest_task())

        self.ws_task = self.loop.create_task(self.auth.websocket())
        self.auth.register_ws_callback(self.callback)

    async def close(self):
        """Close the session."""
        for task in [
            self.ws_task,
            self.token_task,
            self.rest_task,
        ]:
            tasks = []
            if task is not None:
                tasks.append(task)
                if not task.cancelled():
                    task.cancel()
        with contextlib.suppress(asyncio.CancelledError):
            await asyncio.gather(*tasks)

    async def get_status(self) -> MowerList:
        """Get mower status via Rest."""
        mower_list = await self.auth.get_json(AutomowerEndpoint.mowers)
        for idx, _ent in enumerate(mower_list["data"]):
            mower_list["data"][idx]["attributes"].update(
                mower_list["data"][idx]["attributes"]["settings"]
            )
            del mower_list["data"][idx]["attributes"]["settings"]
        self.data = mower_list
        self.mower_as_dict_dataclass()
        return self.mowers

    async def resume_schedule(self, mower_id: str):
        """Resume schedule.

        Remove any ovveride on the Planner and let the mower
        resume to the schedule set by the Calendar.
        """
        body = {"data": {"type": "ResumeSchedule"}}
        url = AutomowerEndpoint.actions.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def pause_mowing(self, mower_id: str):
        """Send pause mowing command to the mower via Rest."""
        body = {"data": {"type": "Pause"}}
        url = AutomowerEndpoint.actions.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def park_until_next_schedule(self, mower_id: str):
        """Send park until next schedule command to the mower."""
        body = {"data": {"type": "ParkUntilNextSchedule"}}
        url = AutomowerEndpoint.actions.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def park_until_further_notice(self, mower_id: str):
        """Send park until further notice command to the mower."""
        body = {"data": {"type": "ParkUntilFurtherNotice"}}
        url = AutomowerEndpoint.actions.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def park_for(self, mower_id: str, duration_in_min: int):
        """Parks the mower for a period of minutes.

        The mower will drive to
        the charching station and park for the duration set by the command.
        """
        body = {
            "data": {
                "type": "Park",
                "attributes": {"duration": duration_in_min},
            }
        }
        url = AutomowerEndpoint.actions.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def start_for(self, mower_id: str, duration_in_min: int):
        """Start the mower for a period of minutes."""
        body = {
            "data": {
                "type": "Park",
                "attributes": {"duration": duration_in_min},
            }
        }
        url = AutomowerEndpoint.actions.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def set_cutting_height(self, mower_id: str, cutting_height: int):
        """Start the mower for a period of minutes."""
        body = {
            "data": {
                "type": "settings",
                "attributes": {"cuttingHeight": cutting_height},
            }
        }
        url = AutomowerEndpoint.settings.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def set_headlight_mode(
        self,
        mower_id: str,
        headlight_mode: Literal[
            HeadlightModes.ALWAYS_OFF,
            HeadlightModes.ALWAYS_ON,
            HeadlightModes.EVENING_AND_NIGHT,
            HeadlightModes.EVENING_ONLY,
        ],
    ):
        """Send headlight mode to the mower."""
        body = {
            "data": {
                "type": "settings",
                "attributes": {"headlight": {"mode": headlight_mode}},
            }
        }
        url = AutomowerEndpoint.settings.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def set_calendar(
        self,
        mower_id: str,
        task_list: list,
    ):
        """Send calendar task to the mower."""
        body = {
            "data": {
                "type": "calendar",
                "attributes": {"tasks": task_list},
            }
        }
        url = AutomowerEndpoint.calendar.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def switch_stay_out_zone(
        self, mower_id: str, stay_out_zone_id: str, switch: bool
    ):
        """Enable or disable a stay out zone."""
        body = {
            "data": {
                "type": "stayOutZone",
                "id": stay_out_zone_id,
                "attributes": {"enable": switch},
            }
        }
        url = AutomowerEndpoint.stay_out_zones.format(
            mower_id=mower_id, stay_out_id=stay_out_zone_id
        )
        await self.auth.patch_json(url, json=body)

    async def _token_monitor_task(self):
        while True:
            if "expires_at" in self.token:
                expires_at = self.token["expires_at"]
                sleep_time = max(MIN_SLEEP_TIME, expires_at - time.time() - MARGIN_TIME)
            else:
                sleep_time = MIN_SLEEP_TIME

            _LOGGER.debug("token_monitor_task sleeping for %s sec", sleep_time)
            await asyncio.sleep(sleep_time)
            await self.auth.async_get_access_token()

    def _update_data(self, j):
        if self.data is None:
            raise NoDataAvailableException
        if self.data is not None:
            for datum in self.data["data"]:
                if datum["type"] == "mower" and datum["id"] == j["id"]:
                    if j["type"] == "positions-event":
                        last_pos_identical = (
                            datum["attributes"]["positions"][0]
                            == j["attributes"]["positions"][0]
                        )
                        if not last_pos_identical:
                            j["attributes"]["positions"].extend(
                                datum["attributes"]["positions"]
                            )
                    for attrib in j["attributes"]:
                        try:
                            tasks = j["attributes"]["calendar"]["tasks"]
                            if len(tasks) == 0:
                                temp_task = datum["attributes"]["calendar"]["tasks"]
                                datum["attributes"][attrib] = j["attributes"][attrib]
                                datum["attributes"]["calendar"]["tasks"] = temp_task
                            if len(tasks) > 0:
                                datum["attributes"][attrib] = j["attributes"][attrib]
                        except KeyError:
                            datum["attributes"][attrib] = j["attributes"][attrib]
        self.mower_as_dict_dataclass()
        self._schedule_data_callbacks()

    def mower_as_dict_dataclass(self):
        """Convert mower data to a dictionary DataClass."""
        mowers_list = MowerList(**self.data)
        for mower in mowers_list.data:
            self.mowers[mower.id] = mower.attributes

    def callback(self, new_data):
        """Pass received websocket data to the update function."""
        if new_data:
            self._update_data(new_data)

    async def _rest_task(self):
        """Poll data periodically via Rest."""
        while True:
            await asyncio.sleep(REST_POLL_CYCLE)
            self.data = await self.get_status()
            self._schedule_data_callbacks()
