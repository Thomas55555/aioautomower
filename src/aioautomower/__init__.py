"""Automower library using aiohttp."""

__all__ = [
    "auth",
    "const",
    "exceptions",
    "model",
    "session",
    "utils",
]
