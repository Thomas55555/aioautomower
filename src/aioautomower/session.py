"""Module to connect to Automower with websocket."""

import asyncio
import contextlib
import datetime
import logging
from dataclasses import dataclass
from typing import Any, Iterable, Literal, Mapping

from aiohttp import WSMessage, WSMsgType

from .auth import AbstractAuth
from .const import EVENT_TYPES, REST_POLL_CYCLE
from .exceptions import NoDataAvailableException, TimeoutException
from .model import HeadlightModes, MowerAttributes
from .utils import mower_list_to_dictionary_dataclass

_LOGGER = logging.getLogger(__name__)

logging.basicConfig(level=logging.DEBUG)


@dataclass
class AutomowerEndpoint:
    """Endpoint URLs for the AutomowerConnect API."""

    mowers = "mowers/"
    "List data for all mowers linked to a user."

    actions = "mowers/{mower_id}/actions"
    "Accepts actions to control a mower linked to a user."

    calendar = "mowers/{mower_id}/calendar"
    "Update the calendar on the mower."

    settings = "mowers/{mower_id}/settings"
    "Update the settings on the mower."

    stay_out_zones = "mowers/{mower_id}/stayOutZones/{stay_out_id}"
    "Enable or disable the stay-out zone."

    work_area_cutting_height = "mowers/{mower_id}/workAreas/{work_area_id}"
    "This will update cutting height on the work area."

    work_area_calendar = "mowers/{mower_id}/workAreas/{work_area_id}/calendar"
    "Update the calendar for a work area on the mower."

    error_confirm = "mowers/{mower_id}/errors/confirm"
    "Confirm mower non-fatal error"


class _MowerCommands:
    """Sending commands."""

    def __init__(
        self,
        auth: AbstractAuth,
    ):
        """Send all commands to the API.

        :param class auth: The AbstractAuth class from aioautomower.auth.
        """
        self.auth = auth

    async def resume_schedule(self, mower_id: str):
        """Resume schedule.

        Remove any override on the Planner and let the mower
        resume to the schedule set by the Calendar.
        """
        body = {"data": {"type": "ResumeSchedule"}}
        url = AutomowerEndpoint.actions.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def pause_mowing(self, mower_id: str):
        """Send pause mowing command to the mower via Rest."""
        body = {"data": {"type": "Pause"}}
        url = AutomowerEndpoint.actions.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def park_until_next_schedule(self, mower_id: str):
        """Send park until next schedule command to the mower."""
        body = {"data": {"type": "ParkUntilNextSchedule"}}
        url = AutomowerEndpoint.actions.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def park_until_further_notice(self, mower_id: str):
        """Send park until further notice command to the mower."""
        body = {"data": {"type": "ParkUntilFurtherNotice"}}
        url = AutomowerEndpoint.actions.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def park_for(self, mower_id: str, duration_in_min: int):
        """Parks the mower for a period of minutes.

        The mower will drive to
        the charching station and park for the duration set by the command.
        """
        body = {
            "data": {
                "type": "Park",
                "attributes": {"duration": duration_in_min},
            }
        }
        url = AutomowerEndpoint.actions.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def start_in_workarea(
        self,
        mower_id: str,
        work_area_id: int,
        duration_in_min: int | None = None,
    ):
        """Start the mower in a work area for a period of minutes.

        If duration is skipped the mower will continue forever.
        """
        body = {
            "data": {
                "type": "StartInWorkArea",
                "attributes": {"duration": duration_in_min, "workAreaId": work_area_id},
            }
        }
        url = AutomowerEndpoint.actions.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def start_for(self, mower_id: str, duration_in_min: int):
        """Start the mower for a period of minutes."""
        body = {
            "data": {
                "type": "Start",
                "attributes": {"duration": duration_in_min},
            }
        }
        url = AutomowerEndpoint.actions.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def set_cutting_height(self, mower_id: str, cutting_height: int):
        """Set the cutting height for the mower."""
        body = {
            "data": {
                "type": "settings",
                "attributes": {"cuttingHeight": cutting_height},
            }
        }
        url = AutomowerEndpoint.settings.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def set_cutting_height_workarea(
        self, mower_id: str, cutting_height: int, work_area_id: int
    ):
        """Set the cutting height for a specific work area."""
        body = {
            "data": {
                "type": "workArea",
                "id": work_area_id,
                "attributes": {"cuttingHeight": cutting_height},
            }
        }
        url = AutomowerEndpoint.work_area_cutting_height.format(
            mower_id=mower_id, work_area_id=work_area_id
        )
        await self.auth.patch_json(url, json=body)

    async def set_headlight_mode(
        self,
        mower_id: str,
        headlight_mode: Literal[
            HeadlightModes.ALWAYS_OFF,
            HeadlightModes.ALWAYS_ON,
            HeadlightModes.EVENING_AND_NIGHT,
            HeadlightModes.EVENING_ONLY,
        ],
    ):
        """Send headlight mode to the mower."""
        body = {
            "data": {
                "type": "settings",
                "attributes": {"headlight": {"mode": headlight_mode}},
            }
        }
        url = AutomowerEndpoint.settings.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def set_calendar(
        self,
        mower_id: str,
        task_list: list[str],
    ):
        """Send calendar task to the mower."""
        body = {
            "data": {
                "type": "calendar",
                "attributes": {"tasks": task_list},
            }
        }
        url = AutomowerEndpoint.calendar.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)

    async def switch_stay_out_zone(
        self, mower_id: str, stay_out_zone_id: str, switch: bool
    ):
        """Enable or disable a stay out zone."""
        body = {
            "data": {
                "type": "stayOutZone",
                "id": stay_out_zone_id,
                "attributes": {"enable": switch},
            }
        }
        url = AutomowerEndpoint.stay_out_zones.format(
            mower_id=mower_id, stay_out_id=stay_out_zone_id
        )
        await self.auth.patch_json(url, json=body)

    async def error_confirm(self, mower_id: str):
        """Confirm non-fatal mower error."""
        body = {}  # type: dict[str, str]
        url = AutomowerEndpoint.error_confirm.format(mower_id=mower_id)
        await self.auth.post_json(url, json=body)


class AutomowerSession:
    """Automower API to communicate with an Automower.

    The `AutomowerSession` is the primary API service for this library. It supports
    operations like getting a status or sending commands.
    """

    def __init__(
        self,
        auth: AbstractAuth,
        poll: bool = False,
    ) -> None:
        """Create a session.

        :param class auth: The AbstractAuth class from aioautomower.auth.
        :param bool poll: Poll data with rest if True.
        """
        self._data: dict[str, Iterable[Any]] | None = {}
        self.auth = auth
        self.commands = _MowerCommands(self.auth)
        self.pong_cbs: list = []
        self.data_update_cbs: list = []
        self.data: dict[str, MowerAttributes] = {}
        self.last_ws_message: datetime.datetime
        self.loop: asyncio.AbstractEventLoop = asyncio.get_running_loop()
        self.poll = poll
        self.rest_task: asyncio.Task | None = None

    def register_data_callback(self, callback) -> None:
        """Register a data update callback."""
        if callback not in self.data_update_cbs:
            self.data_update_cbs.append(callback)

    def _schedule_data_callback(self, cb) -> None:
        """Schedule a data callback."""
        if self.poll and self.data is None:
            raise NoDataAvailableException
        self.loop.call_soon_threadsafe(cb, self.data)

    def _schedule_data_callbacks(self) -> None:
        """Schedule a data callbacks."""
        for cb in self.data_update_cbs:
            self._schedule_data_callback(cb)

    def unregister_data_callback(self, callback) -> None:
        """Unregister a data update callback.

        :param func callback: Takes one function, which should be unregistered.
        """
        if callback in self.data_update_cbs:
            self.data_update_cbs.remove(callback)

    def register_pong_callback(self, pong_callback) -> None:
        """Register a pong callback.

        It's not real ping/pong, but a way to check if the websocket
        is still alive, by receiving an empty message.
        """
        if pong_callback not in self.pong_cbs:
            self.pong_cbs.append(pong_callback)

    def _schedule_pong_callback(self, cb) -> None:
        """Schedule a pong callback."""
        self.loop.call_soon_threadsafe(cb, self.last_ws_message)

    def _schedule_pong_callbacks(self) -> None:
        """Schedule pong callbacks."""
        for cb in self.pong_cbs:
            self._schedule_pong_callback(cb)

    def unregister_pong_callback(self, pong_callback) -> None:
        """Unregister a pong update callback.

        :param func callback: Takes one function, which should be unregistered.
        """
        if pong_callback in self.pong_cbs:
            self.pong_cbs.remove(pong_callback)

    async def connect(self) -> None:
        """Connect to the API.

        This method handles the login. Also a REST task will be started, which
        periodically polls the REST endpoint, when polling is set to true.
        """
        self._schedule_data_callbacks()

        if self.poll:
            await self.get_status()
            self.rest_task = asyncio.create_task(self._rest_task())

    def add_settigs_tree(self, msg_dict: dict) -> dict:
        """Add settings key and ignore empty calendar tasks.

        Needed, because when polling the API `headlight` and `cuttingHeight` are
        nested under the `settings` key. But when receiving a websocket message,
        the `settings` key is missing.
        """
        copy_msg_dict = dict(msg_dict)
        current_data = self._data
        if current_data is None:
            raise NoDataAvailableException
        dater_iter = current_data["data"]
        for _, current_data_mower in enumerate(dater_iter):
            if current_data_mower["id"] == copy_msg_dict["id"]:
                current_attributes = current_data_mower["attributes"]
                formated_msg = {
                    "id": current_data_mower["id"],
                    "type": "settings-event",
                    "attributes": {
                        "calendar": {"tasks": current_attributes["calendar"]["tasks"]},
                        "settings": {
                            "cuttingHeight": current_attributes["settings"][
                                "cuttingHeight"
                            ],
                            "headlight": {
                                "mode": current_attributes["settings"]["headlight"][
                                    "mode"
                                ]
                            },
                        },
                    },
                }
                new_attributes = copy_msg_dict["attributes"]
                if len(new_attributes["calendar"]["tasks"]) > 0:
                    formated_msg["attributes"]["calendar"]["tasks"] = copy_msg_dict[
                        "attributes"
                    ]["calendar"]["tasks"]
                if "cuttingHeight" in new_attributes:
                    formated_msg["attributes"]["settings"]["cuttingHeight"] = (
                        copy_msg_dict["attributes"]["cuttingHeight"]
                    )
                if "headlight" in new_attributes:
                    formated_msg["attributes"]["settings"]["headlight"]["mode"] = (
                        copy_msg_dict["attributes"]["headlight"]["mode"]
                    )
                return formated_msg
        return copy_msg_dict

    def _handle_text_message(self, msg: WSMessage) -> None:
        """Process a text message to data."""
        if not msg.data:
            self.last_ws_message = datetime.datetime.now(tz=datetime.UTC)
            _LOGGER.debug("last_ws_message:%s", self.last_ws_message)
            self._schedule_pong_callbacks()
        if msg.data:
            msg_dict = msg.json()
            if "type" in msg_dict:
                if msg_dict["type"] in EVENT_TYPES:
                    if msg_dict["type"] == "settings-event":
                        copy = dict(msg_dict)
                        msg_dict = self.add_settigs_tree(copy)
                    _LOGGER.debug("Got %s, data: %s", msg_dict["type"], msg_dict)
                    self._update_data(msg_dict)
                else:
                    _LOGGER.warning("Received unknown ws type %s", msg_dict["type"])
            elif "ready" in msg_dict and "connectionId" in msg_dict:
                _LOGGER.debug(
                    "Websocket ready=%s (id='%s')",
                    msg_dict["ready"],
                    msg_dict["connectionId"],
                )

    async def start_listening(self) -> None:
        """Start listening to the websocket (and receive initial state)."""
        while not self.auth.ws.closed:
            try:
                msg = await self.auth.ws.receive(timeout=300)
                if msg.type in (
                    WSMsgType.CLOSE,
                    WSMsgType.CLOSING,
                    WSMsgType.CLOSED,
                ):
                    break
                if msg.type == WSMsgType.TEXT:
                    self._handle_text_message(msg)
                elif msg.type == WSMsgType.ERROR:
                    continue
            except TimeoutError as exc:
                raise TimeoutException from exc

    async def send_empty_message(self) -> None:
        """Send an empty message every 60s."""
        while True:
            await asyncio.sleep(60)
            _LOGGER.debug("ping:%s", datetime.datetime.now(tz=datetime.UTC))
            await self.auth.ws.send_str("")

    async def get_status(self) -> dict[str, MowerAttributes]:
        """Get mower status via Rest."""
        mower_list = await self.auth.get_json(AutomowerEndpoint.mowers)
        self._data = mower_list
        self.data = mower_list_to_dictionary_dataclass(self._data)
        return self.data

    def _update_data(self, new_data) -> None:
        """Update internal data, with new data from websocket."""
        if self._data is None:
            raise NoDataAvailableException

        data = self._data["data"]
        for mower in data:
            if mower["type"] == "mower" and mower["id"] == new_data["id"]:
                new_attributes: Mapping[Any, Any] = new_data["attributes"]
                value: Mapping[Any, Any]
                for attrib, value in new_attributes.items():
                    mower["attributes"][attrib] = value
        self.data = mower_list_to_dictionary_dataclass(self._data)
        self._schedule_data_callbacks()

    async def _rest_task(self) -> None:
        """Poll data periodically via Rest."""
        while True:
            await self.get_status()
            self._schedule_data_callbacks()
            await asyncio.sleep(REST_POLL_CYCLE)

    async def close(self) -> None:
        """Close the session."""
        if self.rest_task:
            if not self.rest_task.cancelled():
                self.rest_task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await asyncio.gather(self.rest_task)
