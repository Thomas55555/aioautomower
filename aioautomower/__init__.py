"""Automower library using aiohttp."""
from .rest import *
from .session import AutomowerSession
