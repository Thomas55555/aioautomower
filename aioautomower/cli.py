#!/usr/bin/env python3
import argparse
import asyncio
import logging
import signal
import time

from aiohttp import ClientSession

from aioautomower.auth import AbstractAuth
from aioautomower.session import AutomowerSession
import aioautomower.utils

from .const import API_BASE_URL

_LOGGER = logging.getLogger(__name__)


async def run_tester(client_id: str, client_secret: str):
    automower_api = AutomowerSession(
        AsyncConfigEntryAuth(ClientSession(), client_id, client_secret), poll=True
    )
    automower_api.register_data_callback(
        lambda x: logging.info("data callback;%s", x),
    )

    await automower_api.connect()

    def sigusr1():
        asyncio.ensure_future(automower_api.invalidate_token())

    def sigusr2():
        asyncio.ensure_future(automower_api.get_status())

    loop = asyncio.get_event_loop()
    loop.add_signal_handler(signal.SIGUSR1, sigusr1)
    loop.add_signal_handler(signal.SIGUSR2, sigusr2)

    while True:
        await asyncio.sleep(0.1)


def main():
    """Tester for the Husqvarna Automower API.

    The tester will login using client_id and client_secret and connect to a
    websocket listening for mower updates.

    The tester listens to two signals on which it performs the following
    requests:
    SIGUSR1: Invalidate token
    SIGUSR2: Get status
    """

    parser = argparse.ArgumentParser(
        description=main.__doc__, formatter_class=argparse.RawTextHelpFormatter
    )
    parser.add_argument(
        "-s", "--client_id", required=True, help="Husqvarna Application key"
    )
    parser.add_argument(
        "-k", "--client_secret", required=True, help="Husqvarna Application secret"
    )

    args = parser.parse_args()

    logging.basicConfig(level="DEBUG", format="%(asctime)s;%(levelname)s;%(message)s")
    asyncio.run(run_tester(args.client_id, args.client_secret))


class AsyncConfigEntryAuth(AbstractAuth):
    """Provide Husqvarna Automower authentication tied to an OAuth2 based config entry."""

    def __init__(
        self,
        websession: ClientSession,
        client_id: str,
        client_secret: str,
    ) -> None:
        """Initialize Husqvarna Automower auth."""
        super().__init__(websession, API_BASE_URL)
        self.client_id = client_id
        self.client_secret = client_secret
        self.token = None
        self.websession = websession

    async def async_get_access_token(self) -> str:
        """Return a valid access token."""
        if not self.token:
            self.token = await aioautomower.utils.async_get_access_token(
                self.client_id, self.client_secret, self.websession
            )
            token_structured = await aioautomower.utils.async_structure_token(
                self.token["access_token"]
            )
            _LOGGER.debug("token_structured.exp: %s", token_structured.exp)
            _LOGGER.debug(time.time())
        if token_structured.exp < time.time():
            self.token = await aioautomower.utils.async_get_access_token(
                self.client_id, self.client_secret, self.websession
            )
        return self.token["access_token"]
