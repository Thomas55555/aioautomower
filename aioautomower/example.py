"""An example file to use this library."""

import asyncio
import logging
import time
from typing import cast

from aiohttp import ClientSession

from aioautomower.auth import AbstractAuth
from aioautomower.const import API_BASE_URL
from aioautomower.session import AutomowerSession
from aioautomower.utils import async_get_access_token, async_structure_token

_LOGGER = logging.getLogger(__name__)


CLIENT_ID = "1e33fa27-ca34-4762-9a9e-5967f873a731"
CLIENT_SECRET = "763adf3c-1b16-4c3b-91cd-c07316243881"
CLOCK_OUT_OF_SYNC_MAX_SEC = 20


class AsyncTokenAuth(AbstractAuth):
    """Provide Husqvarna Automower authentication tied to an OAuth2 based config entry."""

    def __init__(
        self,
        websession: ClientSession,
    ) -> None:
        """Initialize Husqvarna Automower auth."""
        super().__init__(websession, API_BASE_URL)
        self.token = None

    async def async_get_access_token(self) -> str:
        """Return a valid access token."""
        if not self.token:
            self.token = await async_get_access_token(CLIENT_ID, CLIENT_SECRET)
            token_structured = await async_structure_token(self.token["access_token"])
            print("Token expires at: ", token_structured.exp)
        return self.token["access_token"]

    @property
    def valid_token(self) -> bool:
        """Return if token is still valid."""
        return (
            cast(float, self.token["expires_at"])
            > time.time() + CLOCK_OUT_OF_SYNC_MAX_SEC
        )

    async def async_ensure_token_valid(self) -> None:
        """Ensure that the current token is valid."""
        if self.valid_token:
            return

        self.token = await async_get_access_token()


async def main():
    """Establish connection to mower and print states for 5 minutes."""
    websession = ClientSession()
    automower_api = AutomowerSession(AsyncTokenAuth(websession), poll=True)
    # Add a callback, can be done at any point in time and
    # multiple callbacks can be added.
    automower_api.register_data_callback(callback)
    await automower_api.connect()
    await asyncio.sleep(3000)
    # The close() will stop the websocket and the token refresh tasks
    await automower_api.close()


def callback(ws_data):
    """Process websocket callbacks and write them to the DataUpdateCoordinator."""
    print("Mowers data:", ws_data)


asyncio.run(main())
