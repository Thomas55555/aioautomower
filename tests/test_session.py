"""Test automower session."""

import json
from dataclasses import fields
from unittest.mock import AsyncMock

from syrupy.assertion import SnapshotAssertion

from aioautomower.session import AutomowerSession
from tests import load_fixture

MOWER_ID = "c7233734-b219-4287-a173-08e3643f89f0"


async def test_connect(snapshot: SnapshotAssertion):
    """Test automower session."""
    auth_mock = AsyncMock()
    session = AutomowerSession(auth_mock)

    # Call the connect method and assert
    await session.connect()
    auth_mock.get_json.return_value = json.loads(load_fixture("raw_data.json"))
    data = await session.get_status()
    for field in fields(data[MOWER_ID]):
        field_name = field.name
        field_value = getattr(data[MOWER_ID], field_name)
        assert field_value == snapshot(name=f"{field_name}")
