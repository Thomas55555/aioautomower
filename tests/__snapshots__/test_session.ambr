# serializer version: 1
# name: test_connect[battery]
  Battery(battery_percent=100)
# ---
# name: test_connect[calendar]
  Tasks(tasks=[Calendar(start=1140, duration=300, monday=True, tuesday=False, wednesday=True, thursday=False, friday=True, saturday=False, sunday=False, work_area_id=123456), Calendar(start=0, duration=480, monday=False, tuesday=True, wednesday=False, thursday=True, friday=False, saturday=True, sunday=False, work_area_id=0)])
# ---
# name: test_connect[capabilities]
  Capabilities(headlights=True, work_areas=True, position=True, stay_out_zones=True)
# ---
# name: test_connect[cutting_height]
  4
# ---
# name: test_connect[headlight]
  Headlight(mode='EVENING_ONLY')
# ---
# name: test_connect[metadata]
  Metadata(connected=True, status_dateteime=datetime.datetime(2023, 10, 18, 22, 58, 52, 683000, tzinfo=datetime.timezone.utc))
# ---
# name: test_connect[mower]
  Mower(mode='MAIN_AREA', activity='PARKED_IN_CS', state='RESTRICTED', error_code=0, error_key=None, error_datetime=None, error_timestamp=0)
# ---
# name: test_connect[planner]
  Planner(next_start_datetime=datetime.datetime(2023, 6, 5, 19, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200), 'CEST')), next_start_timestamp=1685991600000, override=Override(action='NOT_ACTIVE'), restricted_reason='WEEK_SCHEDULE')
# ---
# name: test_connect[positions]
  list([
    Positions(latitude=35.5402913, longitude=-82.5527055),
    Positions(latitude=35.5407693, longitude=-82.5521503),
    Positions(latitude=35.5403241, longitude=-82.5522924),
    Positions(latitude=35.5406973, longitude=-82.5518579),
    Positions(latitude=35.5404659, longitude=-82.5516567),
    Positions(latitude=35.5406318, longitude=-82.5515709),
    Positions(latitude=35.5402477, longitude=-82.5519437),
    Positions(latitude=35.5403503, longitude=-82.5516889),
    Positions(latitude=35.5401429, longitude=-82.551536),
    Positions(latitude=35.5405489, longitude=-82.5512195),
    Positions(latitude=35.5404005, longitude=-82.5512115),
    Positions(latitude=35.5405969, longitude=-82.551418),
    Positions(latitude=35.5403437, longitude=-82.5523917),
    Positions(latitude=35.5403481, longitude=-82.5520054),
  ])
# ---
# name: test_connect[statistics]
  Statistics(cutting_blade_usage_time=1234, number_of_charging_cycles=1380, number_of_collisions=11396, total_charging_time=4334400, total_cutting_time=4194000, total_drive_distance=1780272, total_running_time=4564800, total_searching_time=370800)
# ---
# name: test_connect[stay_out_zones]
  StayOutZones(dirty=False, zones={'81C6EEA2-D139-4FEA-B134-F22A6B3EA403': Zone(name='Springflowers', enabled=True), 'AAAAAAAA-BBBB-CCCC-DDDD-123456789101': Zone(name='Danger Zone', enabled=False)})
# ---
# name: test_connect[system]
  System(name='Test Mower 1', model='450XH-TEST', serial_number=123)
# ---
# name: test_connect[work_areas]
  dict({
    0: WorkArea(name=None, cutting_height=50),
    123456: WorkArea(name='Front lawn', cutting_height=50),
  })
# ---
